### ISSUE TITLE
[Translation Efficiency Boost] The Magic "Paste as Translation" Operator Proposal

### ISSUE THREAD

**Issue by LeonCheung** [SubtitleEdit/subtitleedit#3246]
A new workflow for translators for a few reasons:

- 1. Not everybody can access online transltor API **easily and freely**, like Google Translate (GT) and Microsoft Translator(MT), because API keys are requested.

- 2. Actually, there are many local sophisticated translation engines that are more professional and focusing on specific language, which can generate even better results than GT and MT. Unfortuntely, it's hardly possible for SE to implement them all.

Luckily, most translation engines support free online translation (e.g. for max length of 5,000). So, I would suggest a new workflow, which I hope can solve many Google-translate-not-work problems:

**Step 1**: Select lines you want to translate, then copy texts only to clipboard (currenlty available as shortcut);

**Step 2**: Open online translation engine such as https://translate.google.com/ or https://www.bing.com/translator, then paste the lines there;

**Step 3**: Copy the translated results to clipcoard, then use the magic **"Paste as Translation"** tool to paste them back to selected lines, one line one subtitle. (It can be done because most translation engines keep line breaks in results)

That's it! Eveything can be done in **10 seconds**!
---

**Benefits:**

- 1. No API keys required.
- 2. No translation engine limit, you can use any engine you want.
- 3. More flexible than current SE translation tools, because: 
  - For the Auto-translate tools in SE (from main menu), you maybe have to translate the whole project at one time;
  - By using the context operator "Google translate selected lines" (from context menu), lines can be selectable, but the results will overwirte the original text, which is no good either for contrast translation or bilingual format.
- 4. You can even paste your own translation instead of machine translation. (as long as line count matches)
- 5. According to statistics, the 5,000 length limit can be generally enough for a 10-minute movie script. So, even for a one hour movie translation, you only need to repeat the copy-paste operation for a few times. Done!

**Comment by Paupertino** [comment]
Rubbish. 1445 lines? The translator take only 79.
Useful future IS GONE because of greed.

**Comment by LeonCheung** [comment]
@Paupertino Sorry but I don't quite follow you here. What userful future is gone? What's the greed on earth?

**Comment by Paupertino** [comment]
Until now, It was very easy to translate subtitles. In whole, and quickly. Using the program. Now- IT IS NOT. Because there is no free API to do this. PAY, MONEY are needed. I think, I'm perfectly clear. Sure, the program is fantastic. And I'm deeply thankful.

**Comment by Paupertino** [comment]
You were very helpful- just the proposed method is now a working approach.

**Comment by LeonCheung** [comment]
SE is already very powerful with different workflows. For example, another way to approach this is to first export plain texts, then translate online in the way I mentioned above, then import plain text back, then import timecodes from original srt. But that's an much less efficient way. 

I don't expect it being considered right now in the coming release, just want to share my humble ideas with devs.

**Comment by darnn** [comment]
Yeah, I wanna say, on the few occasions I needed to translate subtitles, I didn't even remember the Google Translate option was in SE, and I just did it the manual way described above, and it was fine. I get that it's probably different for someone who uses it every day, but what can you do.

**Comment by Dinodo** [comment]
I put this idea in another thread and then I saw it was closed. I'm newbie and only SE user. Sorry for my bad English. I'd like to suggest an alternative. Would it be possible to use embedded translation in Excel to aid in the translation of subtitles?

**Comment by LeonCheung** [comment]
I think the pain spot is that B and G can not be ideal for every SE user and every language. There should be a universal way to open the gate a bit.

**Comment by niksedk** [comment]
Could be a nice backup for GT/BT + other translation engines... but should probably be a plugin.

**Comment by niksedk** [comment]
How does the plugin "Copy-paste translate" work?

**Comment by OmrSi** [comment]
@niksedk Wow! That's some awesome stuff.
I tested on Arabic but it showed the Arabic letters as question marks.
![screenshot 71](https://user-images.githubusercontent.com/20923700/50736064-c8b1cf80-11c0-11e9-8604-7a5560452269.png)

**Comment by LeonCheung** [comment]
Thanks so much Nikolaj, I tried that but just feel a bit confusing on the workflow. It seems we have to translate, in this way,  the whole file instead of selection, right? Sorry about the previouis long proposal, I really should have made a GIF to show it easy to understand. I just took some time to make a GIF below to fully show the intention, as well as the flexibility, since I tried to switch between engines in this demo. (GIF may last for 40 secs around, but the real operation would be very fast I believe):

![new](https://user-images.githubusercontent.com/494788/50737999-1bfc4000-120a-11e9-8f60-358d3f281cbe.gif)

In this way, the total chars to be copied to clip board could be easily and visiually controlled via solution like this:
![44](https://user-images.githubusercontent.com/494788/50738112-83ff5600-120b-11e9-92fe-ebd12266199f.png)

Either way, I need to highly appreciate for your great endeavor on it!

**Comment by LeonCheung** [comment]
Btw, I may mess up with the concept between "Original" and "Transtaion"  in the demo, but you may get the point anway. :)

**Comment by niksedk** [comment]
I think the plugin is easier... as you don't have to think about size and copying to clipboard is done automatically... also, translate for selected lines is kinda fixed in latest beta:

![image](https://user-images.githubusercontent.com/1261331/50855604-95ba3800-1388-11e9-8656-9b1fd69d6ced.png)

**Comment by LeonCheung** [comment]
Terrific! Now I know the thing puzzled me was kinda bug, the initial release just kept on copying only the first block for unkown reason, instead of the continous ones, even after pasting back. Now everything becomes clear. And true, it's quite convenient for translating the whole files. Yeah, you are so great!~

Just one thing, if you click the cancel button, there's no way to continue the work but to start it over from the first line. Do you think we need a solution here?

**Comment by niksedk** [comment]
@LeonCheung: thx :)
Version 0.4 now starts from first selected line.

**Comment by LeonCheung** [comment]
Working like a charm. Thx!

**Comment by Paupertino** [comment]
The plugin works, but put ALL lines in ONE BIG single line. Sorry.

**Comment by LeonCheung** [comment]
@Paupertino assuming you need a guid for the basic workflow:

1. Open the plugsin from **Menu > Auto-translate > Translate via copy/pate...**.
2. Use a Use a proper seperator, like "|", or, in many cases, simply use none. Leave **Max Text Size** as 5000 by default, since that's the limit for most online translation engine. Then click **Translate**. 
3. Click "Copy source text to clippoard" button (or **Ctrl C**), paste it in the textbox in any engine, then copy the result back to clipboard.
4. Go back to SE and press the big button (or **Ctrl V**) to paste result from clipboard. 

If the total word count is larger than 5,000, just keep the panel open and repeat step 3 and 4 until every thing finished. 

Basically, only **five clicks** needed per 5,000 words. Isn't that acceptable for you?
